firmware/examples/FactoryTest/source/main.cpp

```diff
@@ -0,0 +1,323 @@
+}
+
+void OnBoardLedTest(bool gpio_test, bool i2c_test, bool flash_test)
+{
+  // Turn on all LEDs if all tests pass else none.
+  // LED Test
+  OnBoardLed leds;
+  leds.Initialize();
+  if (gpio_test && i2c_test && flash_test)
+  {
+    leds.SetAll(0x0F);
+  }
+  LOG_INFO("Test Results:\n\tGPIO:%d\n\tI2C:%d\n\tFlash:%d\n\t", gpio_test,
+           i2c_test, flash_test);
+  Delay(1000);
+}
+
+bool GestureSensorTest()
+{
+  // Gesture Sensor Test
+  printf("++++++++++++++++++++++++++++++++++++++\n\n");
+  printf("Starting Gesture Sensor Test...\n\n");
+  uint8_t kApds9960IdRegisterAddress = 0x92;
+  uint8_t apds9960_id                = 0x00;
+  bool result                        = false;
+  // uint8_t control_sequence[] = { 0x8F, 0x00 };
+  // i2c.WriteThenRead(kGestureAddress, &control_sequence[0], 1,
+  //                   &control_sequence[0], 1);
+  // LOG_INFO("Control 1 = 0x%02X", control_sequence[0]);
+  // control_sequence[0] = 0x90;
+  // i2c.WriteThenRead(kGestureAddress, &control_sequence[0], 1,
+  //                   &control_sequence[0], 1);
+  // LOG_INFO("Control 2 = 0x%02X", control_sequence[0]);
+  // control_sequence[0] = 0x9F;
+  // i2c.Write(kGestureAddress, control_sequence, sizeof(control_sequence));
+
+  uint8_t initialization_sequence[] = { 0x80, 0x05 };
+  i2c.Write(kGestureAddress, initialization_sequence,
+            sizeof(initialization_sequence));
+
+  i2c.WriteThenRead(kGestureAddress, &kApds9960IdRegisterAddress, 1,
+                    &apds9960_id, 1);
+
+  LOG_INFO("Gesture ID = 0x%02X", apds9960_id);
+  LOG_INFO("End of Gesture Sensor Test...");
+
+  result = CheckDeviceId(apds9960_id, 0xAB);
+
+  return result;
+}
+
+bool AccelerometerTest()
+{
+  // Accelerometer Test
+  printf("++++++++++++++++++++++++++++++++++++++\n\n");
+  printf("Starting Accelerometer Test...\n\n");
+  uint8_t kMMAIdRegisterAddress     = 0x0D;
+  uint8_t mma_id                    = 0x00;
+  uint8_t initialization_sequence[] = { 0x2A, 0x01 };
+  bool result                       = false;
+
+  i2c.Write(kAccelerometerAddress, initialization_sequence,
+            sizeof(initialization_sequence));
+
+  i2c.WriteThenRead(kAccelerometerAddress, &kMMAIdRegisterAddress, 1, &mma_id,
+                    1);
+
+  LOG_INFO("Accelerometer ID = 0x%02X", mma_id);
+  LOG_INFO("End of Accelerometer Test...");
+
+  result = CheckDeviceId(mma_id, 0x2A);
+
+  return result;
+}
+
+bool TemperatureSensorTest()
+{
+  // Temperature Test
+  // ID should be 0x01 [7:4]
+  printf("++++++++++++++++++++++++++++++++++++++\n\n");
+  printf("Starting Temperature Test...\n\n");
+  uint8_t kTemperatureIdAddress = 0xC0;
+  uint8_t temperature_id        = 0x00;
+  bool result                   = false;
+
+  i2c.WriteThenRead(kTemperatureAddress, &kTemperatureIdAddress, 1,
+                    &temperature_id, 1);
+  LOG_INFO("ID:%02X", temperature_id);
+  result = CheckDeviceId(temperature_id, 0x14);
+
+  uint8_t control_sequence[] = { 0xC4, 0x00, 0xC9, 0x01 };
```

Comment:
Create a helper function for calculating temperature. Pass an i2c reference to it.